src/core/algorithms/nd/nd_verifier/util/stats_calculator.cpp

```diff
@@ -0,0 +1,99 @@
+#include "algorithms/nd/nd_verifier/util/stats_calculator.h"
+
+#include <memory>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+#include <easylogging++.h>
+
+#include "algorithms/nd/nd.h"
+
+namespace algos::nd_verifier::util {
+
+template <typename It>
+std::shared_ptr<std::vector<size_t>> StatsCalculator::CalculateFrequencies(size_t codes_number,
+                                                                           It begin, It end) {
+    auto result = std::make_shared<std::vector<size_t>>();
+    for (size_t code{0}; code < codes_number; ++code) {
+        result->push_back(std::count(begin, end, code));
+    }
+
+    return result;
+}
+
+[[nodiscard]] std::unordered_map<std::string, size_t> StatsCalculator::GetLhsFrequencies() const {
+    if (lhs_values_->size() != lhs_frequencies_->size()) {
+        throw std::runtime_error(
+                "(StatsCalulator): lhs_frequencies contains an incorrect number of values");
+    }
+
+    std::unordered_map<std::string, size_t> result;
+    for (size_t i{0}; i < lhs_values_->size(); ++i) {
+        result.emplace((*lhs_values_)[i], (*lhs_frequencies_)[i]);
+    }
+
+    return result;
+}
+
+[[nodiscard]] std::unordered_map<std::string, size_t> StatsCalculator::GetRhsFrequencies() const {
+    if (rhs_values_->size() != rhs_frequencies_->size()) {
+        throw std::runtime_error(
+                "(StatsCalulator): rhs_frequencies contains an incorrect number of values");
+    }
+
+    std::unordered_map<std::string, size_t> result;
+    for (size_t i{0}; i < rhs_values_->size(); ++i) {
+        result.emplace((*rhs_values_)[i], (*rhs_frequencies_)[i]);
+    }
+
+    return result;
+}
+
+void StatsCalculator::CalculateStats() {
+    lhs_frequencies_ =
+            CalculateFrequencies(lhs_values_->size(), encoded_lhs_->begin(), encoded_lhs_->end());
+    rhs_frequencies_ =
+            CalculateFrequencies(rhs_values_->size(), encoded_rhs_->begin(), encoded_rhs_->end());
+
+    std::vector<size_t> highlights_lhs_codes;
+    model::WeightType max_weight{0};
+    model::WeightType min_weight{UINT_MAX};
+    for (size_t code{0}; code < lhs_values_->size(); ++code) {
+        size_t weight = value_deps_[code].size();
+
+        if (weight == 0) {
+            // This value is associated with no rhs values, but is somehow present in value_deps
+            LOG(INFO) << "WARNING: Lhs value '" << (*lhs_values_)[code]
+                      << "''s weight is 0, but it's present in value_deps";
+        }
+
+        if (weight > max_weight) {
+            max_weight = weight;
+            highlights_lhs_codes.clear();
+            highlights_lhs_codes.push_back(code);
+        } else if (weight == max_weight) {
+            highlights_lhs_codes.push_back(code);
```

Comment:
I see a possible large number of unnecessary operations. For example, consider sequence [1,...,1,2,...,2,...,n,...,n]. You will perform a lot of operations (`clear()`, `push_back()`), but only codes that correspond to the maximum weight are important. So we should avoid unnecessary operations. For example, we can calculate minimum and maximum (with their indices) and only then update `highlights_lhs_codes`. If possible, you can try to use `std::minmax_element()` or `std::max_element`.